### ISSUE TITLE
Odd issue sending/receiving messages

### ISSUE THREAD

**Issue by elad** [hidespb/node-whatsapi#43]
Hello,

I'm experiencing two odd issues with node-whatsapi. I wrote a little echo "server" that's supposed to receive messages and send them back. It's a simple sequence where I `connect`, `login`, `requestServerProperties`, `sendIsOnline`, and then wait for `receivedMessage` events, calling `sendMessage` with the relevant parts.

When I run the code, it connects and logs in fine. It receives an incoming message:

```
rx <message from="xxx@s.whatsapp.net" id="1429991181-130" type="text" t="1429994279" notify="Elad">
rx   <body>6 data length</body>
rx </message>

tx <receipt to="xxx@s.whatsapp.net" type="read" id="1429991181-130" t="1429994280"></receipt>
Incoming message: { body: 'Test 2',
  from: 'xxx@s.whatsapp.net',
  author: '',
  id: '1429991181-130',
  date: Sat Apr 25 2015 23:37:59 GMT+0300 (IDT),
  notify: 'Elad',
  isGroup: false }

tx <message to="xxx@s.whatsapp.net" type="text" id="message-1429994280-2" t="1429994280">
tx   <body>6 data length</body>
tx </message>
```

But apparently doesn't send read receipts because I only see one tick in the sending client. I also don't see the target client online.

When sending messages, I never receive the response from the server.

Then, I quit the app (^C it) and start it again. At that point I see two blue ticks in the sending client, and this in the app:

```
rx <ack from="xxx@s.whatsapp.net" id="1429991181-130" class="receipt" type="read"></ack>
```

The messages still don't get sent though.

What am I doing wrong? :)

**Comment by matteocontrini** [comment]
I have no idea TBH, but this seems to be related to #41, where only the first sent message gets actually delivered. There's probably something blocking the sockets channel at some point.

I will investigate when I have some time... Thanks for the report

**Comment by elad** [comment]
Great, thanks, and please - let me know if there's anything I can do to help!

**Comment by matteocontrini** [comment]
Can you try what's described [here](https://github.com/mgp25/Chat-API/issues/625)?
When the process gets stuck, leave the script running, delete the conversation from the client app and see what happens, if something happens...

**Comment by elad** [comment]
Tried it, doesn't seem to help. Here's an isolated version of the code I'm using:

```
var whatsapi = require('./node-whatsapi/whatsapi.js');

var wa = whatsapi.createAdapter({
    msisdn: 'xxx',
    username: 'xxx',
    password: 'xxx',
    ccode: 'xxx'
}, true);

wa.connect(function (err) {
  if (err) {
    throw err;
  }

  wa.login(function (err) {
    if (err) {
      throw err;
    }

    wa.sendIsOnline();

    wa.on('receivedMessage', function (message) {
      console.log(message);
    });
  });
});
```

If I understand correctly, read receipts are automatically sent back, so all messages sent to the app should be read immediately. Unfortunately that's not the case and I don't see any ticks with this version... regardless of whether I quit the app and start it again. I see incoming messages printed, but the sending client (my phone) doesn't see the other end online and all messages have only one tick.

**Comment by matteocontrini** [comment]
I can't reproduce the problem with your code.
Are receipts sent back from what you can see from the debug?

**Comment by elad** [comment]
Looks like it:

```
tx <receipt to="xxx@s.whatsapp.net" type="read" id="1429991181-165" t="1429996997"></receipt>
```

**Comment by CODeRUS** [comment]
you first need to send normal receipt and THEN send read receipt to mark message as received in sender client

**Comment by matteocontrini** [comment]
That's what I was trying right now, thank you for advice! :+1:

**Comment by elad** [comment]
@CODeRUS, @matteocontrini: How do I do that? I assumed read receipts are handled internally by the library. What should I add to make it work the way you describe?

**Comment by matteocontrini** [comment]
I'll push a fix for that later :)

**Comment by CODeRUS** [comment]
@elad i'm sorry, but i never used this library and don't know how it works, i just know how every part of whatsapp api works, you can ask me if you need something. i just entered this thread by some reference link from another projects.

**Comment by elad** [comment]
@matteocontrini thanks :)

@CODeRUS at the moment I'm just a library consumer, but thanks for your input in solving this problem!

**Comment by matteocontrini** [comment]
@CODeRUS this library is basically a port of the PHP library, so the structure is similar but implemented with the fast Node.js. I am keeping this project active with the help of some collaborators, and now we need to fix these little bugs...

**Comment by CODeRUS** [comment]
@matteocontrini it's okay, i'm just fan of c++ and doing my own library for mobile/desktop clients :)

**Comment by matteocontrini** [comment]
@elad can you try with the latest commit?
I think I weren't able to reproduce the problem because I was testing with the Windows Phone app, which is pretty buggy on this aspect (e.g. the app itself sends the read receipt *BEFORE* the normal receipt).

I will add an option to disable automatic receipts (one or both of them).

**Comment by elad** [comment]
@matteocontrini sure, in ~90 minutes when I get home, I don't have my password with me at the moment. Will follow up as soon as I try.

**Comment by matteocontrini** [comment]
@CODeRUS where is the PHP library sending the two receipts? I can see only one call, for the `read` one, [here](https://github.com/mgp25/Chat-API/blob/master/src/whatsprot.class.php#L2635)

**Comment by elad** [comment]
@matteocontrini just updated to ccb4f25135d7b3b76a5c2365ec430c99debf5dfe and it still doesn't work.

I'm sending a message from my phone to the node-whatsapi client, I see the incoming message, I see the outgoing receipts:

```
tx <receipt to="xxx@s.whatsapp.net" id="1430070425-25" t="1430071734"></receipt>

tx <receipt to="xxx@s.whatsapp.net" id="1430070425-25" t="1430071734" type="read"></receipt>
```

But still only one tick. If I ^C the app and start it again I get two ticks.

I should note that I still don't see the node-whatsapi client as "online" even though it calls `sendIsOnline()`.

Any ideas? :)

**Comment by elad** [comment]
Data point: After a couple of minutes I get this on the console:

```
rx <stream:error>
rx   <ack></ack>
rx </stream:error>
```

Searching around (and reading posts from @CODeRUS) I get the impression that this is because the client isn't sending an `ack` for either a `message` or `notification`. Looking at the code I don't see any notifications, so I'm guessing this has to do with messaging.

Is it possible there's a `this.sendNode(this.createAckNode(node));` missing somewhere?

**Comment by matteocontrini** [comment]
Can you enable the debug mode and post the logs? So we can see what is missing...

**Comment by matteocontrini** [comment]
Can you post the full debug log of what happens after successful login?
And can you try sending messages with another phone (or WhatsApp installation), because I can't reproduce the problem and there must be something wrong with your specific device, since it's not receiving anything from node-whatsapi.

And... what's the mobile device you're testing with?

**Comment by elad** [comment]
Hi,

I'm testing with an iPhone 4. I'm using the above code. Here's the full log:

```
$ node rx

tx <stream:features>
tx   <readreceipts></readreceipts>
tx   <groups_v2></groups_v2>
tx   <privacy></privacy>
tx   <presence></presence>
tx </stream:features>

tx <auth mechanism="WAUTH-2" user="xxx">106 data length</auth>

rx <start from="s.whatsapp.net"></start>

rx <stream:features></stream:features>

rx <challenge>20 data length</challenge>

tx <response xmlns="urn:ietf:params:xml:ns:xmpp-sasl">36 data length</response>

rx <success t="1430073085" props="4" kind="free" status="active" creation="1429973310" expiration="1461509310">20 data length</success>

tx <presence name="yyy"></presence>

rx <ib from="s.whatsapp.net">
rx   <dirty type="groups" timestamp="1429973310"></dirty>
rx </ib>

rx <message from="zzz@s.whatsapp.net" id="1430070425-29" retry="1" offline="1" type="text" t="1430072067" notify="Elad">
rx   <body>4 data length</body>
rx </message>

tx <receipt to="zzz@s.whatsapp.net" id="1430070425-29" t="1430073087"></receipt>

tx <receipt to="zzz@s.whatsapp.net" id="1430070425-29" t="1430073087" type="read"></receipt>
{ body: 'Word',
  from: 'zzz@s.whatsapp.net',
  author: '',
  id: '1430070425-29',
  date: Sun Apr 26 2015 21:14:27 GMT+0300 (IDT),
  notify: 'Elad',
  isGroup: false }

rx <ib from="s.whatsapp.net">
rx   <offline count="1"></offline>
rx </ib>

rx <message from="zzz@s.whatsapp.net" id="1430070425-55" type="text" t="1430073114" notify="Elad">
rx   <body>4 data length</body>
rx </message>

tx <receipt to="zzz@s.whatsapp.net" id="1430070425-55" t="1430073116"></receipt>

tx <receipt to="zzz@s.whatsapp.net" id="1430070425-55" t="1430073116" type="read"></receipt>
{ body: 'Test',
  from: 'zzz@s.whatsapp.net',
  author: '',
  id: '1430070425-55',
  date: Sun Apr 26 2015 21:31:54 GMT+0300 (IDT),
  notify: 'Elad',
  isGroup: false }

rx <stream:error>
rx   <ack></ack>
rx </stream:error>

^C
```

Is it possible my specific device/number has been temporarily blocked/banned...? I barely sent messages with it and only to numbers in contacts. I also receive messages so I tend to think that's not it, especially given two ticks after restart...

**Comment by matteocontrini** [comment]
I'm not seeing any `ack` of the receipts, are you really seeing the two twicks in the app? Can you try with another phone perhaps? Or send me the number via email and I can try that.

**Comment by elad** [comment]
That's actually a session where I don't get two ticks... If you're not seeing any `ack` maybe that's the problem then? :)

**Comment by matteocontrini** [comment]
This is how receipts-acks work. Client A is your phone, Client B is node-whatsapi.
* Client A sends the message
* Client B receives the message
* Client B sends a receipt to Client A, telling "hey, I got/read your message"
* Client A gets the receipt and sends an `ack`nowledge, so that client B knows that everything went fine

The problem here is that client A, your phone, isn't sending any ack. The stream error means that you received a message but you haven't sent any receipt. And WhatsApp thinks this because you don't get any ack of that.

Plus, the fact that you don't see the Client B online is really strange, and makes me think that the problem is related with your Client A (the phone).

Again, can you try with another phone? With phone I mean another Client A.

**Comment by elad** [comment]
I see. Unfortunately not at the moment because if Whatsapp *is* blocking me I can't afford that on my real phone. :/

**Comment by matteocontrini** [comment]
What do you mean? WhatsApp isn't blocking you. What makes you think that?

**Comment by elad** [comment]
I haven't had a message from the app arrive my phone since yesterday. Everything else worked in the broken fashion I described earlier...

**Comment by CODeRUS** [comment]
@matteocontrini you need to send ack after receiving receipts. It's what whatsapp server is trying to tell you :)

**Comment by matteocontrini** [comment]
There's no incoming receipt, look at the logs.

**Comment by elad** [comment]
Okay, to test this out further I signed into Whatsapp with the official app and did the registration again. Now things seem to work fine with the exact same code, so I'm happy. :) I'll report back if things break again. Thanks!

**Comment by CODeRUS** [comment]
@matteocontrini okay, but i see ib dirty instead, maybe it's the reason. You need to send clean dirty iq node.
```
<iq id=ID type="set" to="s.whatsapp.net", xmlns="urn:xmpp:whatsapp:dirty">
    <clean type="groups" />
</iq>
```

**Comment by matteocontrini** [comment]
It used to do that, but probably when we were still on WAUTH-1.
Anyway, what's the meaning of that clean node?

**Comment by CODeRUS** [comment]
dirty generates when you doing something wrong with api, it increasing internal whatsapp counter marking your account as suspect, helping them to identify it as unofficial.

**Comment by matteocontrini** [comment]
I'm stuck in a similar situation. I do get messages, but when I reply and get a receipt, `ack`s don't seem to do its job.
```
rx  <receipt from="xxx@s.whatsapp.net" id="message-1430425216-3" retry="3" offline="3" type="read" t="1430425216"></receipt>

tx  <ack to="xxx@s.whatsapp.net" id="message-1430425216-3" type="read" t="1430426321"></ack>

rx  <receipt from="xxx@s.whatsapp.net" id="message-1430425216-3" offline="0" t="1430425216"></receipt>

tx  <ack to="xxx@s.whatsapp.net" id="message-1430425216-3" type="read" t="1430426321"></ack>

[....]

rx  <ib from="s.whatsapp.net">
rx    <offline count="8"></offline>
rx  </ib>
```

This debug log is using this repository, but it is *identical* when using the PHP version of WhatsAPI.
Acks are being sent, but apparently ignored by the server, so the `retry` count goes up and they even duplicate on script restart. After `retry` reaches 4, the receipt goes away.

Of course leaving the script running will cause a stream error.

I've found similar reports on Chat-API.
* https://github.com/mgp25/Chat-API/issues/658
* https://github.com/mgp25/Chat-API/issues/661 (see the `retry`?)

@CODeRUS any ideas? And I suspect reading the "manual" wouldn't help in this case :P

**Comment by CODeRUS** [comment]
your acks doesn't seems to be right
message receipt ack:
```
<ack class="receipt" type=(receipt type or "delivery") id=(receipt id) />
```
notification ack:
```
<ack to=(notification from) class="notification" id=(notification id) type=(notification type) participant=(notification participant, only if set) from=(notification to, only if set)>
    <sync contacts="out" /> (node is set only if notification type is contacts)
</ack>
```
call ack:
```
<ack to=(call from) class="call" id=(call id) type=(call node children tag) />
```

**Comment by matteocontrini** [comment]
Is this something that changed recently? Anyway it doesn't work, same result and this is the log. Is it right?

```
rx <receipt from="xxx@s.whatsapp.net" id="message-1430425216-3" retry="2" offline="2" t="1430425216"></receipt>

tx <ack class="receipt" id="message-1430425216-3" type="delivery"></ack>

rx <receipt from="xxx@s.whatsapp.net" id="message-1430425219-4" offline="0" type="read" t="1430425218"></receipt>

tx <ack class="receipt" id="message-1430425219-4" type="read"></ack>
```

**Comment by CODeRUS** [comment]
You sending message yourself? Your message id should not contain any prefixes. It should be in time-sequence form.

**Comment by matteocontrini** [comment]
I'm sending messages to another phone number. I removed the prefix but nothing changes. I ported the `ack` modification to Chat-API (locally) but it doesn't work there either.
The log is always the same: successful login, I get the receipts and they are acked, then the offline count is received.

If you want to help and have an idea of what to try, I can send you the phone number and password in private.

**Comment by CODeRUS** [comment]
you can add me in skype: icoderus

**Comment by matteocontrini** [comment]
Fixed with this commit: e144a1ab0cb643ed53968060c86ec862e9c9af54

Many thanks to @CODeRUS :+1:

**Comment by elad** [comment]
Thanks to both of you, @CODeRUS & @matteocontrini! :)
